### PR TITLE
Add support for setting default value to the configuration

### LINKED ISSUE #695: Support set default value to the configuration
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**Description**:

Currently, we dependent on YAML's default value for the configuration of plugin, for example , we have to say "disableJobOrder" because YAML's default vaule of bool is false :( 

One solution is to set the member of `PluginConf` to be a pointer and have a setDefaultFunc for default values.

### PR DESCRIPTION
**Pull request by hex108**
**What this PR does / why we need it**:
Add support for setting default value to the configuration

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #695 

**Special notes for your reviewer**:

**Release note**:

```release-note
Add support for setting default value to the configuration
```

### CODE REVIEW

**Comment by hex108** [comment]
@k82cn The test is happy now. PTAL.

File: pkg/scheduler/actions/allocate/allocate.go [hunk 1]
```diff
@@ -153,7 +157,9 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
 				if err != nil {
 					glog.V(3).Infof("Error in Calculating Priority for the node:%v", err)
 				} else {
+					workerLock.Lock()
```
Comment (k82cn) [inline pkg/scheduler/actions/allocate/allocate.go]: let's creat a seperate PR for that.

File: pkg/scheduler/framework/session_plugins.go [hunk 2]
```diff
@@ -66,7 +66,7 @@ func (ssn *Session) Reclaimable(reclaimer *api.TaskInfo, reclaimees []*api.TaskI
 
 	for _, tier := range ssn.Tiers {
 		for _, plugin := range tier.Plugins {
-			if plugin.ReclaimableDisabled {
+			if plugin.EnabledReclaimable == nil || !*plugin.EnabledReclaimable {
```
Comment (k82cn) [inline pkg/scheduler/framework/session_plugins.go]: add help func for them.

File: pkg/scheduler/plugins/conformance/conformance.go [hunk 3]
```diff
@@ -21,9 +21,24 @@ import (
 	"k8s.io/kubernetes/pkg/apis/scheduling"
 
 	"github.com/kubernetes-sigs/kube-batch/pkg/scheduler/api"
+	"github.com/kubernetes-sigs/kube-batch/pkg/scheduler/conf"
 	"github.com/kubernetes-sigs/kube-batch/pkg/scheduler/framework"
 )
 
+const name = "conformance"
+
+func init() {
```
Comment (k82cn) [inline pkg/scheduler/plugins/conformance/conformance.go]: please do not modify plugins one by one, that's ok to set default directly after we load config.

File: pkg/scheduler/framework/plugins.go [hunk 4]
```diff
@@ -47,6 +51,39 @@ func GetPluginBuilder(name string) (PluginBuilder, bool) {
 	return pb, found
 }
 
+var pluginConfMutex sync.Mutex
```
Comment (k82cn) [inline pkg/scheduler/framework/plugins.go]: I means just set default value of each config to make it simple :)

File: pkg/scheduler/actions/allocate/allocate.go (hunk 1 above)
Comment (hex108) [reply pkg/scheduler/actions/allocate/allocate.go]: OK. Create #700

File: pkg/scheduler/plugins/conformance/conformance.go (hunk 3 above)
Comment (hex108) [reply pkg/scheduler/plugins/conformance/conformance.go]: Sorry, I do not catch it. How to set default directly?

File: pkg/scheduler/util.go [hunk 5]
```diff
@@ -51,6 +51,8 @@ func loadSchedulerConf(confStr string) ([]framework.Action, []conf.Tier, error)
 	if err := yaml.Unmarshal(buf, schedulerConf); err != nil {
 		return nil, nil, err
 	}
+	framework.ApplyPluginConfDefaultFuncs(schedulerConf)
+
```
Comment (k82cn) [inline pkg/scheduler/util.go:62]: just set default value of plugins one by one here, or provide a defaultFunc here.

File: pkg/scheduler/framework/session_plugins.go (hunk 2 above)
Comment (hex108) [reply pkg/scheduler/framework/session_plugins.go]: Addressed

**Comment by hex108** [comment]
@k82cn All problems addressed. PTAL.

File: pkg/scheduler/plugins/defaults.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+package plugins
+
+import "github.com/kubernetes-sigs/kube-batch/pkg/scheduler/conf"
+
+// ApplyPluginConfDefaults sets option's filed to its default value if not set
+func ApplyPluginConfDefaults(option *conf.PluginOption) {
+	t := true
+
+	switch option.Name {
+	case "conformance":
```
Comment (k82cn) [inline pkg/scheduler/plugins/defaults.go]: It's not necessary to check plugins, just set `true` as default value :)

Comment (hex108) [reply pkg/scheduler/plugins/defaults.go]: Finally get it now! :)

**Comment by hex108** [comment]
@k82cn Addressed. PTAL.

**Review by k82cn — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**